File: android/app/src/main/java/kr/personal/oscalert/Repo.java
package kr.personal.oscalert;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * The market snapshot both tabs show. market.json is fetched from the repository's release
 * asset and kept on disk; live quotes are layered on top when the user refreshes.
 */
final class Repo {
    private Repo() {}

    interface Listener {
        void onChanged();
    }

    static volatile List<Stock> stocks = Collections.emptyList();
    static volatile String asof = "", error = "";
    static volatile long quotesAt = 0;
    static volatile boolean loading = false;

    private static final List<Listener> LISTENERS = new CopyOnWriteArrayList<>();
    private static final ExecutorService IO = Executors.newSingleThreadExecutor();
    private static final Handler MAIN = new Handler(Looper.getMainLooper());

    static void listen(Listener l) { LISTENERS.add(l); }

    static void unlisten(Listener l) { LISTENERS.remove(l); }

    private static void changed() {
        MAIN.post(() -> { for (Listener l : LISTENERS) l.onChanged(); });
    }

    static File file(Context c) {
        return new File(c.getFilesDir(), "market.json");
    }

    /** Fetches market.json and stores it. Returns the parsed stocks. Blocking. */
    static List<Stock> download(Context c) throws Exception {
        String text = Net.get(BuildConfig.MARKET_URL + "?t=" + System.currentTimeMillis());
        List<Stock> parsed = parse(text);          // validate before overwriting the cache
        try (FileOutputStream out = new FileOutputStream(file(c))) {
            out.write(text.getBytes(StandardCharsets.UTF_8));
        }
        return parsed;
    }

    static List<Stock> parse(String text) throws Exception {
        JSONObject root = new JSONObject(text);
        JSONArray rows = root.getJSONArray("stocks");
        List<Stock> out = new ArrayList<>(rows.length());
        for (int i = 0; i < rows.length(); i++) out.add(Stock.parse(rows.getJSONObject(i)));
        asof = root.optString("asof");
        return out;
    }

    /** Loads the cached file if nothing is in memory yet. */
    static void ensure(Context c) {
        if (!stocks.isEmpty() || loading) return;
        File f = file(c);
        if (f.exists()) {
            IO.execute(() -> {
                try {
                    stocks = parse(new String(Files.readAllBytes(f.toPath()), StandardCharsets.UTF_8));
                    changed();
                } catch (Exception ignored) {
                }
            });
        }
        refresh(c, true);
    }

    /**
     * Re-downloads market.json and then live quotes. `allQuotes` fetches every stock's price
     * (stocks tab); otherwise only the stocks that currently signal (summary tab).
     */
    static void refresh(Context context, boolean allQuotes) {
        Context c = context.getApplicationContext();
        if (loading) return;
        loading = true;
        changed();
        IO.execute(() -> {
            String err = "";
            try {
                List<Stock> fresh = download(c);
                stocks = fresh;
                changed();
                if (allQuotes) {
                    Live.apply(fresh, Live.all());
                } else {
                    List<String> codes = new ArrayList<>();
                    for (List<Stock> l : Signals.group(c, fresh).values())
                        for (Stock s : l) codes.add(s.ticker);
                    if (!codes.isEmpty()) Live.apply(fresh, Live.some(codes));
                }
                quotesAt = System.currentTimeMillis();
            } catch (Exception e) {
                err = e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage();
            }
            error = err;
            loading = false;
            changed();
        });
    }
}

File: app/lib/core/repo.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:path_provider/path_provider.dart';

import 'bars.dart';
import 'fmt.dart';
import 'holdings.dart';
import 'live.dart';
import 'market_index.dart';
import 'net.dart';
import 'signals.dart' as signals;
import 'stock.dart';

const repoName = String.fromEnvironment('REPO', defaultValue: 'spilak0203-debug/osc-alert');
/// `--dart-define=MARKET_URL=...` points a local build at a test snapshot.
const marketUrl = String.fromEnvironment('MARKET_URL',
    defaultValue: 'https://github.com/$repoName/releases/download/market-data/market-v2.json');

/// The market snapshot every screen shows. market.json is fetched from the repository's release
/// asset and kept on disk; live quotes and the two indices are layered on top on refresh.
class Repo extends ChangeNotifier {
  Repo._();

  static final Repo I = Repo._();

  List<Stock> stocks = const [];
  List<MarketIndex> indices = const [];
  String asof = '', error = '';
  DateTime? quotesAt;
  bool loading = false;

  /// Set once the stocks tab has been opened: from then on every refresh fetches all prices.
  bool wantAll = false;

  /// Tells every screen to redraw, e.g. after a settings change.
  void changed() => notifyListeners();

  static Future<File> file() async => File('${(await getApplicationSupportDirectory()).path}/market.json');

  /// Fetches market.json and stores it. Returns the parsed stocks.
  Future<List<Stock>> download() async {
    final text = await Net.get('$marketUrl?t=${DateTime.now().millisecondsSinceEpoch}');
    final parsed = parse(text); // validate before overwriting the cache
    await (await file()).writeAsString(text);
    return parsed;
  }

  List<Stock> parse(String text) {
    final root = jsonDecode(text) as Map<String, dynamic>;
    final rows = root['stocks'] as List;
    final out = [for (final r in rows) Stock.parse(r as Map<String, dynamic>)];
    asof = '${root['asof'] ?? ''}';
    return out;
  }

  /// Loads the cached file if nothing is in memory yet, then refreshes.
  Future<void> ensure() async {
    if (stocks.isNotEmpty || loading) return;
    try {
      final f = await file();
      if (await f.exists()) {
        stocks = parse(await f.readAsString());
        changed();
      }
    } catch (_) {}
    await refresh(false);
  }

  /// Re-downloads market.json, the indices and live quotes. `allQuotes` fetches every stock's
  /// price (stocks tab); otherwise only the stocks that currently signal (dashboard) and the
  /// user's holdings.
  Future<void> refresh(bool allQuotes) async {
    if (loading) return;
    loading = true;
    changed();
    var err = '';
    try {
      final fresh = await download();
      stocks = fresh;
      Bars.clear();
      changed();
      final idx = <MarketIndex>[];
      for (final code in MarketIndex.codes) {
        try {
          idx.add(await MarketIndex.load(code));
        } catch (_) {}
      }
      indices = idx;
      changed();
      if (allQuotes || wantAll) {
        Live.apply(fresh, await Live.all());
      } else {
        final codes = Holdings.tickers();
        for (final l in signals.group(fresh).values) {
          for (final s in l) {
            codes.add(s.ticker);
          }
        }
        if (codes.isNotEmpty) Live.apply(fresh, await Live.some(codes.toList()));
      }
      quotesAt = DateTime.now();
    } catch (e) {
      err = '$e';
    }
    error = err;
    loading = false;
    changed();
  }

  /// Title for the top bar: which close the signals are from, and what the market is doing.
  String title() {
    if (asof.isEmpty) return loading ? '불러오는 중…' : '신호 없음';
    return '${asof.substring(0, 4)}.${asof.substring(5, 7)}.${asof.substring(8)} 종가 기준';
  }

  String subtitle() {
    final b = StringBuffer();
    if (indices.isNotEmpty) {
      final now = seoulNow();
      final s = indices.first.session(now);
      if (s.isNotEmpty) b.write(s);
      // After a holiday the latest signals are from the last trading day — say so.
      if (s.contains('휴장')) b.write(' · 가장 최근 거래일 신호');
      // The scan runs around 15:50–16:40; until then the latest signals are from the day before.
      if ((s == '장 마감' || s == '장중') && asof != seoulDate(now)) b.write(' · 오늘 신호는 장 마감 후');
    }
    final at = quotesAt;
    if (at != null) {
      if (b.isNotEmpty) b.write(' · ');
      b.write('시세 ${two(at.hour)}:${two(at.minute)}');
    }
    if (loading) b.write(b.isNotEmpty ? ' · 갱신 중' : '갱신 중');
    return b.toString();
  }
}
